### PR TITLE
feat: allow recursion

### PR DESCRIPTION
**Pull request by ajewellamz**
*Description of changes:*

Given a perfectly legal recursive smithy model, smithy-java can find itself in an infinite loop.
Prune DependentShapesWithPaths to avoid this.

Also remove --quatifier-syntax:3, because that blocked me on this change.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java [hunk 1]
```diff
@@ -326,12 +340,16 @@ public static Set<List<ShapeId>> findAllDependentShapesWithPaths(
...
         Set<List<ShapeId>> pathsToShapes = new LinkedHashSet<>(new LinkedHashSet<>());
+        Set<List<ShapeId>> pathsSummary = new LinkedHashSet<>(new LinkedHashSet<>());
 
         // Breadth-first search via getDependencyShapeIds
         final Queue<List<ShapeId>> toTraverse = new LinkedList<>(initialShapeIdsAsPaths);
         while (!toTraverse.isEmpty()) {
             final List<ShapeId> currentShapeIdWithPath = toTraverse.remove();
-            if (pathsToShapes.add(currentShapeIdWithPath)) {
+
+	    // to avoid cycles, only keep the first list with a given first and last element
```
Comment (robin-aws) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: I may be missing some context, but if the BFS is implemented correctly it shouldn't traverse cycles, and it looks like it has the traversed set already. So why is this necessary?

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java (hunk 1 above)
Comment (ajewellamz) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: I can guarantee that cycles are traversed, and you get an arbitrarily large set of arbitrarily large lists.

I'm pretty sure that the problem is with
dependencyShapeIdsWithPaths.forEach(toTraverse::add);
where given (A,B,C) for every dependent X of C we add the list (A,B,C,X) and so we add (A,B,C,B) and (A,B,C,B,C) et cetera.

Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: Ah got it, `toTraverse` is a set of paths rather than nodes, so it doesn't prevent traversing cycles.

Edit: more accurately it's the fact that `pathsToShapes` is a set of paths.

Comment (robin-aws) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: You can make this simpler and clearer by replacing `pathsSummary` by a `Set<ShapeId> visitedShapes` instead. All paths this method produces will always start from the same ShapeID so that's exactly equivalent, and to boot it matches the classic BFS algorithm so it'll be easier to understand.

**Review by robin-aws — APPROVED** [review]
Thanks!

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java (hunk 1 above)
Comment (ajewellamz) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java]: Done

**Review by texastony — COMMENTED** [review]
Comment: We probably need to refactor Smithy-Dafny to use the Knowledge index instead of our various model exploring utility methods.

https://smithy.io/javadoc/1.5.1/software/amazon/smithy/model/knowledge/KnowledgeIndex.html
